keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1504 @@
+resource types. Note that neither of the existing QoS-class resource types (RDT
+or blockio) support updates because of runtime limitations, yet.
+
+```diff
+ message UpdateContainerResourcesRequest {
+
+ ...
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++    // Configuration of class resources.
++    ContainerClassResources class_resources = 5;
+}
+```
+
+The `PodSandboxConfig` will be supplemented with a corresponding
+`class_resources` field that will be the Pod level configuration. Depending on
+the resource this might be interpreted as a pod-level default (that is used if
+nothing is specified in the `ContainerConfig`) or as a true Pod-level setting -
+in the end the detailed behavior will be responsibility of the container
+runtime.
+
+```diff
+ message PodSandboxConfig {
+@@ -45,5 +45,14 @@ message PodSandboxConfig {
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++    // Configuration of QoS-class resources.
++    PodClassResources class_resources = 10;
++
+ }
+
++// PodClassResources specifies the configuration of QoS-class resources
++// resources of a pod.
++message PodClassResources {
++    // QoS-class resource assignment of the pod.
++    // Key is the resource type and values is the class name within the resource type.
++    map<string, string> class = 1;
++}
+```
+
+Also, define "known" QoS-class resource types to more easily align container
+runtime implementations:
+
+```
++
++const (
++       // ClassResourceRdt is the name of the RDT QoS-class resource
++       ClassResourceRdt = "rdt"
++       // ClassResourceBlockio is the name of the blockio QoS-class resource
++       ClassResourceBlockio = "blockio"
++)
+```
+
+### Pod annotations
+
+Use Pod annotation as the initial K8s user interface, similar to e.g. how
+seccomp support was added. This will bridge the gap between enabling
+QoS-class resources in the CRI protocol and making them available in the Pod
+spec.
+
+A feature gate ClassResources enables kubelet to look for pod
+annotations and set the QoS-class resource assignment via CRI protocol accordingly.
+
+Specifically, kubelet will support annotations for specifying RDT and blockio
+class, the two types of QoS-class resources that already have basic support in the
+container runtimes.
+
+  class for all containers
+- `rdt.resources.beta.kubernetes.io/default` for setting a Pod-level default RDT
+  class for all containers
+- `rdt.resources.beta.kubernetes.io/container.<container-name>` for
+  container-specific RDT class settings
+  blockio class for all containers
+- `blockio.resources.beta.kubernetes.io/default` for setting a Pod-level default
+  blockio class for all containers
+- `blockio.resources.beta.kubernetes.io/container.<container-name>` for
+  container-specific blockio class settings
```

Comment:
I've also seen these annotation names suggested:

- `rdt.resources.beta.kubernetes.io/pod` 
- `rdt.resources.beta.kubernetes.io/container.<container-name>` 
- `blockio.resources.beta.kubernetes.io/pod`
- `blockio.resources.beta.kubernetes.io/container.<container-name>`

1. There's a lot of cardinality or whatever here. Any control plane measure to index annotations by key is liable to have a lot of different keys to think about.
2. I thought we didn't put the `beta` bit in the key any more.
3. How come these are beta? The KEP has yet to reach alpha.

Reply:
Will mark this as resolved as annoations have been dropped from the KEP